pkg/apps/metrics/prometheus/metrics.go

```diff
@@ -0,0 +1,142 @@
+)
+
+const (
+	completeRolloutCount         = "complete_rollouts_total"
+	activeRolloutDurationSeconds = "active_rollouts_duration_seconds"
+	lastFailedRolloutTime        = "last_failed_rollout_time"
+
+	availablePhase = "available"
+	failedPhase    = "failed"
+	cancelledPhase = "cancelled"
+)
+
+var (
+	nameToQuery = func(name string) string {
+		return strings.Join([]string{"openshift_apps_deploymentconfigs", name}, "_")
+	}
+
+	completeRolloutCountDesc = prometheus.NewDesc(
+		nameToQuery(completeRolloutCount),
+		"Counts total complete rollouts",
+		[]string{"phase"}, nil,
+	)
+
+	lastFailedRolloutTimeDesc = prometheus.NewDesc(
+		nameToQuery(lastFailedRolloutTime),
+		"Tracks the time of last failure rollout per deployment config",
+		[]string{"namespace", "name", "generation"}, nil,
+	)
+
+	activeRolloutDurationSecondsDesc = prometheus.NewDesc(
+		nameToQuery(activeRolloutDurationSeconds),
+		"Tracks the active rollout duration in seconds",
+		[]string{"namespace", "name", "phase", "generation"}, nil,
+	)
+
+	apps       = appsCollector{}
+	registered = false
+)
+
+type appsCollector struct {
+	lister kcorelisters.ReplicationControllerLister
+}
+
+func IntializeMetricsCollector(rcLister kcorelisters.ReplicationControllerLister) {
+	apps.lister = rcLister
+	if !registered {
+		prometheus.MustRegister(&apps)
+		registered = true
+	}
+	glog.V(4).Info("apps metrics registered with prometheus")
+}
+
+func (c *appsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- completeRolloutCountDesc
+	ch <- activeRolloutDurationSecondsDesc
+}
+
+type failedRollout struct {
+	timestamp  float64
+	generation int64
+}
+
+// Collect implements the prometheus.Collector interface.
+func (c *appsCollector) Collect(ch chan<- prometheus.Metric) {
+	result, err := c.lister.List(labels.Everything())
+	if err != nil {
+		glog.V(4).Infof("Collecting metrics for apps failed: %v", err)
+		return
+	}
+
+	var available, failed, cancelled float64
+
+	latestFailedRollouts := map[string]failedRollout{}
+
+	for _, d := range result {
+		if util.IsTerminatedDeployment(d) {
+			if util.IsDeploymentCancelled(d) {
+				cancelled++
+				continue
+			}
+			if util.IsFailedDeployment(d) {
+				failed++
+				dcName := util.DeploymentConfigNameFor(d)
+
+				// Track the latest failed rollout per deployment config
+				isLatest := true
+				if r, exists := latestFailedRollouts[d.Namespace+"/"+dcName]; exists {
+					if d.Status.ObservedGeneration > r.generation {
+						isLatest = false
+					}
+				} else {
+					isLatest = false
+				}
+				if !isLatest {
+					latestFailedRollouts[d.Namespace+"/"+dcName] = failedRollout{
+						timestamp:  float64(d.CreationTimestamp.Unix()),
+						generation: d.Status.ObservedGeneration,
+					}
+				}
+				continue
+			}
+			if util.IsCompleteDeployment(d) {
+				available++
+				continue
+			}
+		}
+
+		// TODO: possible time screw?
+		durationSeconds := time.Now().Unix() - d.CreationTimestamp.Unix()
```

Comment:
Isn't the time the deployer pod started running more precise than the creation of the RC for when the new rollout started?

Reply:
@kargakis i think there is some time skew here but the deployer pod might be already gone and I will need to get that pod which will slow down the collection?